NFD image compatibility proposal (#1845)
kubernetes-sigs/node-feature-discovery: enhancements/1845-nfd-image-compatibility/README.md

# KEP-1845: Image Compatibility with NFD

## Summary

Today, there is no standard solution of describing container image requirements against hardware or operating systems.

> Cloud-native is being adapted by highly demanded industries where container compatibility plays a critical role to services in terms of cluster preparation and overall performance.

Reviewer: ```suggestion
Cloud-native is being adopted by highly demanded industries where container compatibility plays a critical role for services in terms of cluster preparation and overall performance.
```

This proposal introduces an idea of NFD image compatibility metadata.

> NFD labels can successfully be added to image to describe the requirements against a host or OS.

Reviewer: ```suggestion
NFD labels can successfully be added to the image to describe the requirements against a host or OS.
```

Cloud-native is being adapted by highly demanded industries where container compatibility plays a critical role to services in terms of cluster preparation and overall performance.
This proposal introduces an idea of NFD image compatibility metadata.
NFD labels can successfully be added to image to describe the requirements against a host or OS.

The document has been prepared based on the experience and progress of the [OCI Image Compatibility working group](https://github.com/opencontainers/wg-image-compatibility/tree/main/docs/proposals).

## Motivation

Image compatibility metadata will help container image authors describe compatibility requirements in a standard way.
Metadata will be uploaded with the image to the image registry.
This makes hard container compatibility requirements discoverable, programmable, and will support different consumers and cover use cases where the application requires a specific compatible environment.

### Goals

#### Phase 1

- Use existing NFD labels (keys) to describe container image requirements.
- Create a notation that will describe values including list of ranges and literal.
- Create a new OCI artifact type for compatibility metadata.

> Allow to verify node compatibility - including nodes that are not part of k8s cluster yet.

Reviewer: ```suggestion
- Allow to verify node compatibility - including nodes that are not part of k8s cluster yet.
```

- Make image compatibility a hard requirement for the NFD installation/usage.
- Cover applications ABI compatibility.

## Proposal

Build a new NFD client tool with the following initial scope:

- CRUD OCI artifact.
- Validate nodes based on provided metadata.
- Run directly on a host which is not part of the Kubernetes cluster, or run as a Kubernetes job on a Kubernetes node.

### Design Details

#### OCI Artifact

[An OCI artifact](https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidelines-for-artifact-usage) should be created to store image compatibility metadata on the image side.  
The artifact can be connected with an image over [the subject field](https://github.com/opencontainers/distribution-spec/blob/11b8e3fba7d2d7329513d0cff53058243c334858/spec.md#pushing-manifests-with-subject).

##### Manifest

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "application/vnd.k8s.nfd.image-compatibility.v1",
  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json",
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2
  },
  "layers": [
    {

>       "mediaType": "application/vnd.k8s.nfd.image-compatibility.spec.v1+json",

Reviewer 2: it could be "application/vnd.k8s.nfd.image-compatibility.spec.v1+json" or "application/vnd.k8s.nfd.image-compatibility.spec.v1+yaml" ?  in kubernetes, usually yaml is used.

  }
}
```

##### Artifact Payload (Schema)

- **version** - *string*  
This REQUIRED property specifies the version of the API being used.

- **compatibilities** - *array of object*  
This REQUIRED property is a list of compatibility sets.

  - **rules** - *object*  
  This REQUIRED property is a reference to the spec of [NodeFeatureRule API](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/usage/custom-resources.html#nodefeaturerule).
  The spec makes it possible to describe image requirements using the discovered features from NFD sources.
  For further reading, please review [the documentation](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/usage/customization-guide.html#nodefeaturerule-custom-resource).

  - **weight** - *int*  
  This OPTIONAL property specify the [node affinity weight](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity-weight).

  - **tag** - *string*  
  This OPTIONAL property allows grouping or dividing of compatibility sets.

  - **description** - *string*  
  This OPTIONAL property is intended for a brief description of a compatibility set.

Example

> ```json

Reviewer 2: considering the node feature rules are usually expressed in yaml format, how about use yaml as an example. support both yaml and json is good.

Author: Fixed.

If one image has multiple artifacts, it is up to a client to choose the correct one.
As a default behavior, it is recommended to get the most recent over "created" timestamp.

#### NFD Key Value Notation

NFD key on the image side should accept ranges and literal values. The following notation should be implemented (to describe ranges a standard comparison operators must be used):

- `>, <, <=, >=`
- `string`
- ranges must be separated by semicolon character `;`

Examples:

- `"kernel-version.full": ">=4.15,<5.30"` - greater or equal to 4.15 and lower than 5.30.
- `"kernel-version.full": ">=4.15,<5.25; >5.30"` - greater or equal to 4.15 and lower than 5.25 or greater than 5.30.
- `"kernel-version.full": ">=5.15.x-x-generic"` - great or equal to 5.15 generic kernel flavor.
- `"kernel-version.full": ">=5.15.x-x-generic; >=5.30"` - great or equal to 5.15 generic kernel flavor or greater or equal to any 5.30 kernel version.
- `"kernel-version.full": "5.15"` - must equal to 5.15.

#### NFD client

A new standalone cmd line utility should be implemented for the NFD project that shares the same functionality as [nfd kubectl plugin](https://nfd.sigs.k8s.io/usage/kubectl-plugin).

> Both clients should implemented the following commands:

Author: Maybe it would be better to just have one client?

Reviewer 2: it seems that current nfd client should work as the kubectl plugin, because it has to access k8s api server with credential. 

for compatibility, access to registry may need different credential. And for the sceanrio where the node is not part of k8s cluster, providing k8s credential seems unnecessary.

- `validate` -  validate a NodeFeatureRule object (implemented in kubectl plugin).
- `test` - test a NodeFeatureRule object against a node (implemented in kubectl plugin).

> - `dryrun` - DryRun a NodeFeatureRule object against a NodeFeature file (implemented in kubectl plugin).

Reviewer 3: Can you better explain "dryrun" ? I'm used to seeing it as `--dryrun` to supplement a more clearly obvious action.

Author: I am also used to see it as you. I kept it that way to be compatibile with the [kubectl `dryrun` command](https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/cmd/kubectl-nfd/subcmd/dryrun.go).
